File: 3. Behavioral Design Patterns/Chain of Responsibility/Go/main.go
package main

import "fmt"

// Chain of Responsibility Pattern Example in Go
// Support ticket system with different levels of handlers

type SupportHandler interface {
	SetNext(SupportHandler)
	HandleRequest(*SupportTicket)
}

type SupportTicket struct {
	Issue    string
	Priority int
}

type BaseHandler struct {
	next SupportHandler
}

func (h *BaseHandler) SetNext(handler SupportHandler) {
	h.next = handler
}

type Level1Support struct {
	BaseHandler
}

func (h *Level1Support) HandleRequest(ticket *SupportTicket) {
	if ticket.Priority <= 1 {
		fmt.Printf("Level 1: Handled '%s'\n", ticket.Issue)
	} else if h.next != nil {
		h.next.HandleRequest(ticket)
	} else {
		fmt.Printf("No handler available for '%s'\n", ticket.Issue)
	}
}

type Level2Support struct {
	BaseHandler
}

func (h *Level2Support) HandleRequest(ticket *SupportTicket) {
	if ticket.Priority <= 2 {
		fmt.Printf("Level 2: Handled '%s'\n", ticket.Issue)
	} else if h.next != nil {
		h.next.HandleRequest(ticket)
	} else {
		fmt.Printf("No handler available for '%s'\n", ticket.Issue)
	}
}

type Level3Support struct {
	BaseHandler
}

func (h *Level3Support) HandleRequest(ticket *SupportTicket) {
	fmt.Printf("Level 3: Handled '%s'\n", ticket.Issue)
}

func main() {
	level1 := &Level1Support{}
	level2 := &Level2Support{}
	level3 := &Level3Support{}
	level1.SetNext(level2)
	level2.SetNext(level3)

	level1.HandleRequest(&SupportTicket{"Password reset", 1})
	level1.HandleRequest(&SupportTicket{"Software bug", 2})
	level1.HandleRequest(&SupportTicket{"System crash", 3})
}

File: 3. Behavioral Design Patterns/Chain of Responsibility/Rust/main.rs
// Chain of Responsibility Pattern Example in Rust
// Support ticket system with different levels of handlers

trait SupportHandler {
    fn set_next(&mut self, handler: Box<dyn SupportHandler>);
    fn handle_request(&self, ticket: &SupportTicket);
}

struct SupportTicket {
    issue: String,
    priority: i32,
}

struct Level1Support {
    next: Option<Box<dyn SupportHandler>>,
}

impl Level1Support {
    fn new() -> Self { Self { next: None } }
}

impl SupportHandler for Level1Support {
    fn set_next(&mut self, handler: Box<dyn SupportHandler>) {
        self.next = Some(handler);
    }

    fn handle_request(&self, ticket: &SupportTicket) {
        if ticket.priority <= 1 {
            println!("Level 1: Handled '{}'", ticket.issue);
        } else if let Some(ref next) = self.next {
            next.handle_request(ticket);
        } else {
            println!("No handler available for '{}'", ticket.issue);
        }
    }
}

struct Level2Support {
    next: Option<Box<dyn SupportHandler>>,
}

impl Level2Support {
    fn new() -> Self { Self { next: None } }
}

impl SupportHandler for Level2Support {
    fn set_next(&mut self, handler: Box<dyn SupportHandler>) {
        self.next = Some(handler);
    }

    fn handle_request(&self, ticket: &SupportTicket) {
        if ticket.priority <= 2 {
            println!("Level 2: Handled '{}'", ticket.issue);
        } else if let Some(ref next) = self.next {
            next.handle_request(ticket);
        } else {
            println!("No handler available for '{}'", ticket.issue);
        }
    }
}

struct Level3Support;

impl SupportHandler for Level3Support {
    fn set_next(&mut self, _handler: Box<dyn SupportHandler>) {}

    fn handle_request(&self, ticket: &SupportTicket) {
        println!("Level 3: Handled '{}'", ticket.issue);
    }
}

fn main() {
    let mut level1 = Level1Support::new();
    let mut level2 = Level2Support::new();
    let level3 = Level3Support;

    level2.set_next(Box::new(level3));
    level1.set_next(Box::new(level2));

    level1.handle_request(&SupportTicket { issue: "Password reset".to_string(), priority: 1 });
    level1.handle_request(&SupportTicket { issue: "Software bug".to_string(), priority: 2 });
    level1.handle_request(&SupportTicket { issue: "System crash".to_string(), priority: 3 });
}
